### DISCUSSION TITLE
How is image depth determined for XPM images

### DISCUSSION THREAD

**Discussion by robert-dodier** [ImageMagick/ImageMagick#6921] [Digital Image Processing] [answered]
I'm puzzling over how image depth is determined for XPM images, and towards that end I'm peeking at the code in ImageMagick. 

There doesn't appear to be any place in the image metadata which states the image depth, so I'm guessing that it must be determined by inspecting the strings which specify the color map as hexadecimal strings. Is that how it is carried out in ReadXPMImage? I have searched through xpm.c and other files in ImageMagick (image.h, colormap.h, etc) but I can't figure out where the image depth is actually determined. 

The most pressing question for me is whether or not it is true that image depth for XPM images is not specified and therefore can only be determined by inspecting the color map. Secondarily I would like to see how it is carried out by ImageMagick. Thank you for any light you can shed on this.

**Comment by snibgo** [comment, marked as answer]
The depth is implicit in the colours.

xpm.c function ReadXPMImage() calls QueryColorCompliance() in color.c to set each entry in the palette. When a colour starts with `#`, that function counts the number `n` of hex digits, and sets `depth=4*(n/4)` EDIT or `depth=4*(n/3)`. The depth is returned in the `PixelInfo *color` parameter, and ReadXPMImage() sets the maximum of these to `image->depth`. Other colours have `depth=8`.

**Reply by robert-dodier** [reply]
Thanks, that's very helpful. I will take a look at the code as you mentioned.

I'm a little surprised that the color depth is not part of the XPM image metadata, but that's just me, it's not a problem and I can see how to work with it.
